### ISSUE TITLE
Subsequent az login returns AADSTS50076 for user with MFA 

### ISSUE THREAD

**Issue by alex-zyl** [Azure/azure-cli#20771]
> ### `az feedback` auto-generates most of the information requested below, as of CLI version 2.0.62

**Describe the bug**

*Command*
```az login -u XXX -p XXX```

With az 2.29, the very first call fails with ```interaction_required``` requiring interactive flow via ```az login```. On the second run of ```az login -u XXX -p XXX``` cached subscription list is returned.

After migrating to 2.30 the first call fails with 
```
AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access ''.
To re-authenticate, please run:
az login
``` 
However after performing the interactive login, the second  ```az login -u XXX -p XXX``` run does not return cached data and fails with AADSTS50076.

**To Reproduce**

1. Use user account with enabled MFA.
2. Execute```az login -u XXX -p XXX``` to receive
```
AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access ''.
To re-authenticate, please run:
az login
```
3. Execute ```az login``` to sign in interactively.
4. Execute ```az login -u XXX -p XXX``` again.

**Expected behavior**

Second```az login -u XXX -p XXX``` returns a list of subscriptions and does not issue AADSTS50076.

**Environment summary**

```
Windows 10 20H2
Installer: MSI

"azure-cli": "2.31.0",
"azure-cli-core": "2.31.0",
"azure-cli-telemetry": "1.0.6",
"extensions": {
  "aks-preview": "0.4.71",
  "resource-graph": "1.1.0"
}
```

**Comment by jiasli** [comment]
This is more of a problem of how ADAL and MSAL handle refresh token. 

I think the current behavior of MSAL is expected - if `az login -u XXX -p XXX` is disallowed, it should always be disallowed, regardless of whether interactive `az login` has been done before.

**Comment by rayluo** [comment]
> With az 2.29, the **very first call fails** with interaction_required **requiring interactive flow via az login**. On the second run of az login -u XXX -p XXX cached subscription list is returned.

It is indeed possible that the **very first username password call would fail** and it requires one initial interactive login to "unlock" the username password flow. The reason is about the consent, described in [MSAL Python wiki page](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Username-Password-Authentication):

    Because Username Password Authentication is a non-interactive flow:
    * the user of your application must have previously consented to use the application
      (and that consent would be provided via an interactive log in).

However, if your tenant further requires MFA, then consent along simply won't be enough to "unblock" username password flow. This is also described in that wiki page:

> The Username/Password authentication is not compatible with conditional access and multi-factor authentication, because this is not an interactive flow, the Microsoft Identity platform does not have an opportunity to present a web-based dialog for the end user to interact. As a consequence, **if your app runs in an Azure AD tenant where the tenant admin requires multi-factor authentication (many organizations do that), this flow will not work.**

---

What is your usage scenario, anyway? If you or your end users can do interactive "az login", you can stick with "az login". After the initial login, `az` maintains your current "session", therefore all your subsequent "az ..." commands would all work without interaction.

**Comment by alex-zyl** [comment]
Thanks @rayluo, @jiasli 

Our use case is next: we have a tool that runs a sequence of independent tasks. Each task runs login script as a prerequisite: 
-> Task1: 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; -> login
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; -> business login
-> Task2: 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; -> login
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; -> business logic

The login script simply does ```login -u XXX -p XXX``` and falls back to ```az login``` in case of MFA
With previous CLI version subsequent tasks implicitly relied on interactive flow result made by the very first task, so a user running the tool is prompted with the interactive flow only once. 

With the latest CLI version, a user is currently disturbed with account selection window on each task.

> However, if your tenant further requires MFA, then consent along simply won't be enough to "unblock" username password flow.

So as this behavior is considered as expected, is there any workaround to achieve previous ADAL's behavior other than manually checking current account and token expiration?

**Comment by rayluo** [comment]
> So as this behavior is considered as expected, is there any workaround to achieve previous ADAL's behavior other than manually checking current account and token expiration?

@jiasli is more familiar than me on `az` patterns. Here I'm just brainstorming loud: what if @alex-zyl 's tasks all contain just the "business logic" part? That way, they will all reuse the existing logged-in session, or emit an error/hint to tell user to run `az login` first if they haven't already done so. In a sense, all the built-in `az foo bar` sub-commands work this way.

**Comment by jiasli** [comment]
I tested your scenario, and found **`az login -u XXX -p XXX` command doesn't try to login at all**! ADAL simply returns the refresh token from token cache without any interaction with AAD:

```
cli.azure.cli.core.util: attempting to read file C:\Users\name\.azure\accessTokens.json as utf-8-sig
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - Authority:Performing instance discovery: ...
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - Authority:Performing static instance discovery
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - Authority:Authority validated via static instance discovery
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - TokenRequest:Acquiring token with username password.
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - CacheDriver:finding with query keys: {'_clientId': '...', 'userId': '...'}
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - CacheDriver:Looking for potential cache entries: {'_clientId': '...', 'userId': '...'}
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - CacheDriver:Found 6 potential entries.
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - CacheDriver:Resource specific token found.
adal-python: 74e13fbb-5a74-482c-8e72-3dc834eab42a - CacheDriver:Returning token from cache lookup, AccessTokenId: b'MEipLdncukP2dhFrc91W7T0DtA9Tz6SdC5v1rbUCtSg=', RefreshTokenId: b'tpy2aifJ6jzs2N7WAqxguyuRJ+/yzZH1AFNe6t78W4s='
cli.azure.cli.core.sdk.policies: Request URL: 'https://management.azure.com/tenants?api-version=2019-11-01'
cli.azure.cli.core.sdk.policies: Request method: 'GET'
```

So, even a wrong password will work.

That said, it is now reasonable to just **remove the `az login -u XXX -p XXX` command from your "independent tasks"** as it doesn't take effect at all.

---

Actually, user account is not suitable for automated scenario. If you want to run some task automatically, the recommended approach is to use a service principal: https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli

**Comment by alex-zyl** [comment]
Thanks @yonzhan 
We decided to go this way: on each `login` phase execution we check if there is already login session available via `az account list`. If not, we trigger `az login`.

**Comment by jiasli** [comment]
Great! This is definitely the right approach.
